language/node: multiple improvements

chrmoritz opened the pull request:
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/homebrew/pull/49031).
- [x] Have you successfully run `brew tests` with your changes locally?

-----

This PR contains various little improvements to language/node separated in different commits (each should work independent from the others) for easier review. Please refer to the commit message body for an explanation what the specific commit does.

I've tested a nearly identically language/node version (containing the cache_config, build-from-source and ignore-scripts changes) yesterday with all formulas using language/node in core and their test have passed locally for me.

~~I will need some help fixing the test (see review comments below).~~

chrmoritz commented on Library/Homebrew/test/language/node_spec.rb:
```diff
@@ -22,23 +22,6 @@
-    end
-
-    it "does not raise error with a zero exitstatus" do
-      allow(Utils).to receive(:popen_read).with("npm pack").and_return("pack")
```
How do I mock up 
```
output = `#{pack_cmd}`
```
similar to `output = Utils.popen_read("npm pack").chomp` was mocked up previously?

MikeMcQuaid replied:
```ruby
allow(Kernel).to receive(:`)
```

I think?

chrmoritz replied:
nope, it is 
```
allow(Language::Node).to receive(:`)
```

chrmoritz commented on Library/Homebrew/test/language/node_spec.rb:
```diff
@@ -2,28 +2,14 @@
-      stub_formula_loader(node)
-      allow_any_instance_of(Pathname).to receive(:exist?).and_return(false)
-      expect(ENV).to receive(:prepend_path)
-      subject.setup_npm_environment
```
Looks like `ENV.prepend_path` isn't called in test either, even if I remove the now unneeded line 14.

MikeMcQuaid commented on docs/Node-for-Formula-Authors.md:
````diff
@@ -72,6 +72,8 @@ This will install your Node module in npm's global module style with a custom pr
 bin.install_symlink Dir["#{libexec}/bin/*"]
 ```
 
+**Note:** Because of a required workaround for `npm@5` calling `npm pack` we currently don't support installing modules (from non npm registry tarballs), which requires a prepublish step (e.g. for transpiling sources). See [Homebrew/brew#2820](https://github.com/Homebrew/brew/pull/2820) for more informations.
````
`non-npm`, `which require`, `more information`

MikeMcQuaid commented on docs/Node-for-Formula-Authors.md:
```diff
@@ -47,8 +47,8 @@ Also note that such a formula would only be compatible with the same Node major
-* formulae for standard Node modules compatible with npm's global module format which should use [`std_npm_install_args`](#installing-global-style-modules-with-std_npm_install_args-to-libexec) (like [`azure-cli`](https://github.com/Homebrew/homebrew-core/blob/d93fe9ba3bcc9071b699c8da4e7d733518d3337e/Formula/azure-cli.rb) or [`autocode`](https://github.com/Homebrew/homebrew-core/blob/1a670a6269e1e ...
-* formulae where the `npm install` step is only one of multiple not exclusively Node related install steps (not compatible with npm's global module format) which have to use [`local_npm_install_args`](#installing-module-dependencies-locally-with-local_npm_install_args) (like [`elixirscript`](https://github.com/Homebrew/homebrew-core/blob/ec1e40d37e81af63122a354f0101c377f6a4e66d/Formula/elixirscri ...
+* formulae for standard Node modules compatible with npm's global module format which should use [`std_npm_install_args`](#installing-global-style-modules-with-std_npm_install_args-to-libexec) (like [`azure-cli`](https://github.com/Homebrew/homebrew-core/blob/0f3b27d252b8112c744e0460d871cfe1def6b993/Formula/azure-cli.rb) or [`webpack`](https://github.com/Homebrew/homebrew-core/blob/6282879973d569 ...
+* formulae where the `npm install` step is only one of multiple not exclusively Node related install steps (not compatible with npm's global module format) which have to use [`local_npm_install_args`](#installing-module-dependencies-locally-with-local_npm_install_args) (like [`elixirscript`](https://github.com/Homebrew/homebrew-core/blob/4bb491b7b246830aed57b97348a17e9401374978/Formula/elixirscri ...
```
`only one of multiple not exclusively Node related install steps` is a bit of a mouthful, can you reword? Happy to offer suggestions if you can explain more what this means in a comment

chrmoritz replied:
Just look at the [`elixirscript`](https://github.com/Homebrew/homebrew-core/blob/4bb491b7b246830aed57b97348a17e9401374978/Formula/elixirscript.rb) formula. It can't use `std_npm_install_args` because running npm is not the only needed install step, because it requires several calls to `mix` to build not node related things. The final build is a mix of JavaScript sources and non JavaScript sources.

MikeMcQuaid replied:
I'd maybe just say something like "one of multiple `npm install` calls" or similar?

chrmoritz commented:
I've addressed some doc comments and fixed the `setup_npm_environment` test.

Unfortunately 
```
allow(Kernel).to receive(:`)
```
doesn't seem to work and I still need help from someone with more ruby knowledge to restore this test.

chrmoritz commented:
I've restored/fixed the `std_npm_install_args` test now.

DomT4 commented:
Thanks for looking at this Christian.

chrmoritz commented:
Added one additional `std_npm_install_args` test.

@MikeMcQuaid Could you please take another look?

MikeMcQuaid commented:
Code looks great, just one tiny nit on documentation.

chrmoritz commented:
I've reworded the last documentation nit, but found another remaining issue: Even on a non-verbose `brew install nodeModule` run, the `npm pack` debug output is presented to the user.

MikeMcQuaid commented:
Thanks again @chrmoritz!
